### ISSUE TITLE
PRs cannot be tested in "distribution" form

### ISSUE THREAD

**Issue by clevengr** [pc2ccs/pc2v9#74]
**Describe the issue**: 

When someone submits a PR, reviewers can sometimes "approve" just by looking at the code (changed files).  However, many times the "review" process requires actually running the new code to verify it works correctly.  

The proper way to run the code would be to run a "distribution" -- a fully packaged system built using the same pipeline that will be used when/if the PR is approved for merging into the (develop) code stream.  However, there is no such "distribution" created when a PR is submitted -- so there's no way to really fully test the code.

We need a way to generate a distribution which matches what will be generated when/if the PR is approved.  (One way to do this might be to move to invoking the CI/CD pipeline every time a branch is submitted as a PR, but maybe there's a better way...)

**Comment by lane55** [comment]
Issue #330 PR 821

- goto actions at https://github.com/pc2ccs/pc2v9/actions
- find text "821" on the page
- Click link update old references to (www|pc2).ecs.csus.edu
- Scroll down to Artifacts
- Click on build link to download build.zip

```
build.zip contains
Archive:  build.zip
  Length     EAs   ACLs    Date   Time    Name
 --------    ---   ----    ----   ----    ----
        0      0      0  10/05/23 02:13   api-doc-9.9build-6713~821/
        0      0      0  10/05/23 02:13   pc2-9.9build-6713~821/
  4436906      0      0  10/05/23 02:14   pc2.jar
   333313      0      0  10/05/23 02:13   api-doc-9.9build-6713~821/merge.zip
       33      0      0  10/05/23 02:13   pc2-9.9build-6713~821/merge.tar.gz.md5.txt
       41      0      0  10/05/23 02:13   pc2-9.9build-6713~821/merge.tar.gz.sha1.txt
       65      0      0  10/05/23 02:13   pc2-9.9build-6713~821/merge.tar.gz.sha256.txt
      129      0      0  10/05/23 02:13   pc2-9.9build-6713~821/merge.tar.gz.sha512.txt
 20889393      0      0  10/05/23 02:13   pc2-9.9build-6713~821/merge.tar.gz
       33      0      0  10/05/23 02:14   pc2-9.9build-6713~821/merge.zip.md5.txt
       41      0      0  10/05/23 02:14   pc2-9.9build-6713~821/merge.zip.sha1.txt
       65      0      0  10/05/23 02:14   pc2-9.9build-6713~821/merge.zip.sha256.txt
      129      0      0  10/05/23 02:14   pc2-9.9build-6713~821/merge.zip.sha512.txt
 21843787      0      0  10/05/23 02:13   pc2-9.9build-6713~821/merge.zip
 --------  -----  -----                   -------
 47503935      0      0                   14 files
```

**Comment by lane55** [comment]
Assigned to John to confirm and close this issue.

**Comment by clevengr** [comment]
Although perhaps someone can convince me otherwise, I currently do NOT feel the above "solution" addresses the actual Issue being described, and therefore I'm not prepared to close the issue.

We did talk about this in a recent PC2 meeting. Maybe my understanding at the end of the discussion was wrong -- but I don't believe the suggested approach (i.e., the result of the discussion and also the related comments subsequently added by Doug, above, are the same thing as what is described in this Issue.

Specifically, what Issue #74 is asking for is a way for a Reviewer to say "I'm being asked to test a new PR.  I want a build which will be guaranteed to be byte-for-byte identical to what WOULD BE PRODUCED *IF* this PR which I'm testing were to subsequently be merged into the current state of the DEVELOP branch."  I don't believe the steps we have now allow that -- the "guarantee" is only present *IF THE PR SUBMITTER HAS MERGED DEVELOP INTO THEIR PR BRANCH*. 

Issue #74 is saying that there should be a way for a Reviewer to test EXACTLY the distribution which will result if they Approve the PR and it then immediately gets merged.  In other words, some way to create a "shadow merged distribution"  (not in the sense of Shadow CCS; in the sense of "in the background") which the Reviewer can download and test, and then KNOW that what he's testing is precisely what will get generated when the actual merge to Develop happens.    That is not something that currently happens with our development flow as far as I know; hence the comment in the Issue about possibly changing the CI/CD development flow.

But in any case I don't believe it to be true that the above series of steps will (necessarily) produce a downloadable build that is guaranteed to be identical to what would be built if the PR was Approved and Merged.  Hence, the Issue remains open (although I grant it's somewhat Low Priority...)

**Comment by troy2914** [comment]
If the PR was up to date with the current develop, then the built bits for it would be in effect what would happen if you merged that PR.

**Comment by clevengr** [comment]
> If the PR was up to date with the current develop, then the built bits for it would be in effect what would happen if you merged that PR.

Yep, I'm aware of that; I said as much in my previous comment.  The point is, there is a condition:  _**If** the PR was up to date with the current develop_.  As a Reviewer, I don't want to Approve something because it SEEMS to work; I want to know that it WILL work when it gets merged.  That requires testing against a distribution that is known to be equivalent to what will be produced by a merge -- and yet, we don't want to approve something just to get the merge done just to create the build so we can test it.

There are only two ways to _guarantee_ that the testing done by a Reviewer WILL be the same as what works with a merged distribution:  either have a way to generate exactly what the merged distribution will be (and test that), or Approve the PR, merge it, and then test the resulting distribution.  The latter is fraught with complications because if the results of post-merge testing _fail_, then the merge has to be "backed out" (i.e., the Approval needs to be "revoked").  If there was a way to generate a build that was guaranteed to be the same as what would be the result of Approval/Merge, then the Reviewer could test that -- without having to worry about whether the PR developer took care to make sure that the branch was "up to date with respect to Develop".  (I can point to a number of instances, some of them due to my own mistakes, some due to mistakes of others, where such a condition was NOT true.)

  The Issue is about creating a way to avoid that problem.  And it's NOT the same thing as testing the "Build artifact" -- precisely because there is not (as far as I know) a (reasonable) way to guarantee that the PR Submitter properly pulled Develop into their PR.  

Again, as I've said, this is low priority... but it's something that currently isn't available in our process (and I think it should be... someday...)

**Comment by clevengr** [comment]
Oooo... I just thought of another approach that might solve the problem outlined by this Issue.  Is there perhaps a way to get GitHub to either _outright reject_ an attempt to submit a PR if the PR branch is not up to date with Develop, or to get it to flag the proposed merger as "not up to date" (like it flags PRs that have conflicts)?  Either of those would allow a Reviewer to test the "build artifact" code while being assured that what they were testing matched what would be in the final distribution...

**Comment by troy2914** [comment]
Not sure, gitlab (self hosted) shows when Merge Request is out of date.   Anyways, if you want to be sure,  make sure your develop is in sync with the upstream pc2v9,  get the PR code,  git rebase develop (if that fails, report back to the PR and ask them to rebase it), `ant -f createVERSIONandJar.xml`
then test.
